💃 📦 Add BoxE

ralphabb opened the pull request:
Closes #613 

Following on the issue (https://github.com/pykeen/pykeen/issues/613), I have now written a preliminary version of binary BoxE (which I called BoxE-KG), and am sharing it as is as you suggested. It is based on an interaction (BoxEKGInteraction), and the file with my additions is under src/pykeen/models/unimodal: 
I instantiated boxes analogously to my official code, and set initializations consistently as well. I have also implemented the BoxE distance function and other helpers.

~I have tried to train the model, and this runs. However, I have had some erratic behavior with NSSALoss: When I run this in the forward function (to double-check, this is currently commented-out), I get sensible values. But when I pass this loss to the pipeline, the loss values are minuscule, starting at around 0.001! So I would appreciate some help understanding this. Otherwise, we can proceed from there to verify the model.~ <- This is all solved!!

### Dependencies:

* [x] #623 
* [x] https://github.com/pykeen/pykeen/pull/624
* [x] https://github.com/pykeen/pykeen/pull/626

### Test Code

<details>
<summary>Test code</summary>

```python
import numpy as np
import torch

from pykeen.losses import NSSALoss
from pykeen.datasets import WN18RR
from pykeen.models.unimodal.boxe_kg import BoxE
from pykeen.pipeline import pipeline

from torch.nn import functional

# TODO: Align optimizer settings: Constant LR

class NSSALossLogging(NSSALoss):
    def forward(
        self,
        pos_scores: torch.FloatTensor,
        neg_scores: torch.FloatTensor,
        neg_weights: torch.FloatTensor,
    ) -> torch.FloatTensor:
        # copy of NSSALoss.forward
        neg_loss = functional.logsigmoid(-neg_scores - self.margin)
        neg_loss = neg_weights * neg_loss
        neg_loss = self._reduction_method(neg_loss)
        print("-", -neg_loss.item())

        pos_loss = functional.logsigmoid(self.margin + pos_scores)
        pos_loss = self._reduction_method(pos_loss)
        print("+", -pos_loss.item())

        loss = -pos_loss - neg_loss

        if self._reduction_method is torch.mean:
            loss = loss / 2.0

        return loss



def main():
    embedding_dim = 500
    unif_init_bound = 2 * np.sqrt(embedding_dim)
    init_kw = dict(a=-1 / unif_init_bound, b=1 / unif_init_bound)
    size_init_kw = dict(a=-1, b=1)
    dataset = WN18RR()
    triples_factory = dataset.training
    model = BoxE(
        triples_factory=triples_factory,
        embedding_dim=500,
        norm_order=2,
        tanh_map=True,
        entity_initializer=torch.nn.init.uniform_,
        entity_initializer_kwargs=init_kw,
        relation_initializer=torch.nn.init.uniform_,
        relation_initializer_kwargs=init_kw,
        relation_size_initializer=torch.nn.init.uniform_,
        relation_size_initializer_kwargs=size_init_kw,
    )

    results = pipeline(
        random_seed=1000000,
        dataset=dataset,
        model=model,
        training_kwargs=dict(num_epochs=300, batch_size=512, checkpoint_name="tria.pt", checkpoint_frequency=100),
        loss=NSSALossLogging(margin=5, adversarial_temperature=0.0, reduction="sum"),
        lr_scheduler=torch.optim.lr_scheduler.ConstantLR,
        lr_scheduler_kwargs=dict(total_iters=0),
        training_loop="sLCWA",
        negative_sampler="basic",
        negative_sampler_kwargs=dict(num_negs_per_pos=150),
        result_tracker="json",
        result_tracker_kwargs=dict(name="test.json"),
        evaluation_kwargs=dict(batch_size=16),
        optimizer=torch.optim.Adam,
        optimizer_kwargs=dict(lr=0.001)   # Cancel out the thing
    )

if __name__ == '__main__':
    main()
```
</details>

ralphabb committed:
Preliminary BoxE-KG implementation (NSSALoss yielding weird values, but more sensible when part of forward)

ralphabb committed:
Preliminary BoxE-KG implementation (NSSALoss yielding weird values, but more sensible when part of forward)

cthoyt commented on .github/CODE_OF_CONDUCT.md:
```diff
@@ -59,7 +59,7 @@ further defined and clarified by project maintainers.
 ## Enforcement
 
 Instances of abusive, harassing, or otherwise unacceptable behavior may be
-reported by contacting the [project team.](./project_team.md)
```
not sure these changes are necessary, either way they're irrelevant for this PR so please revert them

ralphabb replied:
Apologies for that, reverted.

cthoyt commented on Tester.py:
```diff
@@ -0,0 +1,30 @@
+                       , training_loop='sLCWA', negative_sampler='basic',  negative_sampler_kwargs=dict(num_negs_per_pos=150),
+                       result_tracker='json', result_tracker_kwargs=dict(name='test.json'),
+                       evaluation_kwargs=dict(batch_size=16), optimizer=torch.optim.Adam)
+    print(results)
```
printing is nice, but what are you actually expecting? code that up as an if/else

ralphabb replied:
I wasn't expecting anything at this point. I simply wanted to show the loss values as part of the PipelineResult. Will remove this.

cthoyt replied:
ahh okay :) I think @mberr will be able to give some advice about the small loss values

cthoyt commented on src/pykeen/models/unimodal/boxe_kg.py:
```diff
@@ -0,0 +1,196 @@
+SANITY_EPSILON = 10 ** -8
+
+
+def product_normalise(input_tensor):   # Geometric normalize
```
needs type annotations and a docstring. what is the purpose of this function? can you write out the math that coresponds to this using latex in the docstring? use triple-double quotes and the sphinx documentation style

```suggestion
def product_normalise(input_tensor: torch.FloatTensor) -> torch.FloatTensor:
```

cthoyt commented on src/pykeen/models/unimodal/boxe_kg.py:
```diff
@@ -0,0 +1,196 @@
+    return pre_norm_out
+
+
+def compute_box(base, delta, size):  # Compute box representations from embeddings
```
need type annotations for all 3 and a docstring

cthoyt commented on src/pykeen/models/unimodal/boxe_kg.py:
```diff
@@ -0,0 +1,196 @@
+    return box_low, box_high
+
+
+def dist(points, box_lows, box_highs):   # The dist function from the paper
```
need type annotations for all inputs and return value, as well as a better name (note that there's a variable inside the function with the same name as the function - hugely problematic)

cthoyt commented on src/pykeen/models/unimodal/boxe_kg.py:
```diff
@@ -0,0 +1,196 @@
+    relation_shape = ('d', 'd', 's', 'd', 'd', 's')  # Boxes are 2xd (size) each, x 2 sets of boxes: head and tail
+    entity_shape = ('d', 'd')   # Base position and bump
+
+    def __init__(self, tanh_map=True, norm_order=2):
```
type annotations and docstring please

ralphabb committed:
Erroneous Typesetting changes reverted

cthoyt commented on src/pykeen/models/unimodal/boxe_kg.py:
```diff
@@ -0,0 +1,196 @@
+        norm_order: int = 2,
+        entity_initializer: Hint[Initializer] = uniform_,
+        entity_initializer_kwargs: Optional[Mapping[str, Any]] = None,
+        relation_initializer: Hint[Initializer] = uniform_,    # Has to be scaled as well
```
there's an alternative function for uniform + norm at `pykeen.nn.init.uniform_norm_`

cthoyt commented on Tester.py:
```diff
@@ -0,0 +1,30 @@
+from src.pykeen.pipeline import pipeline
```
you should be installing pykeen using `pip install -e .` for development mode then just doing normal imports

ralphabb replied:
Thanks for this. Done.

cthoyt committed:
Add unit test for model

cthoyt commented on tests/test_models.py:
```diff
@@ -611,6 +611,12 @@ class TestCrossE(cases.ModelTestCase):
     num_constant_init = 1
 
 
+class TestBoxE(cases.ModelTestCase):
```
@ralphabb you can run this test with `pytest tests/test_models.py::TestBoxE`. Might need a few updates, but this is the general concept we have to make sure that models "work"

ralphabb replied:
I see. I will put this placeholder in for now, and will fill it as I go

cthoyt replied:
As of now this test is passing, so that's a good sign!

ralphabb committed:
Docstrings and annotations added

ralphabb committed:
Fixed quotation marks in test_models

ralphabb committed:
Merge branch 'master' of github.com:ralphabb/pykeen

cthoyt committed:
Apply linting

cthoyt committed:
Cleanup

cthoyt committed:
Reorganize code locations

cthoyt committed:
Clean up math

cthoyt committed:
Update imports

cthoyt committed:
Apply black

cthoyt committed:
Update Tester.py

cthoyt committed:
Delete __init__.py

cthoyt commented on src/pykeen/models/unimodal/boxe_kg.py:
```diff
@@ -0,0 +1,114 @@
+        r"""Initialize BoxE-KG
+
+        :param embedding_dim: The entity embedding dimension $d$. Defaults to 200. Is usually $d \in [50, 300]$.
+        :param tanh_map: Whether to use tanh mapping after BoxE computation. Default - True
```
can you give insight here as to why you would want to use or not use this mapping?

ralphabb replied:
Sure!

ralphabb committed:
Explained tanh map intuition

mberr commented on src/pykeen/utils.py:
```diff
@@ -1095,3 +1100,72 @@ def load_configuration(path: Union[str, pathlib.Path, os.PathLike]) -> Mapping[s
+    log_norm_tensor = torch.log(step2_tensor)  # Compute the log prior to computing the geom. mean
+    step3_tensor = torch.mean(log_norm_tensor, dim=-1, keepdim=True)
+    norm_volume = torch.exp(step3_tensor)
+    pre_norm_out = input_tensor / norm_volume
```
Do we need to care about `norm_volume` being close to zero, i.e., also clamp here?

ralphabb replied:
So the intent here is to make the box volume (whose d^th root is the geometric mean computed above) 1 (and thus geom mean is also 1). I don't think this is the problem, as the official implementation uses this technique as well. But I am open to better variations, of course :)

mberr commented on src/pykeen/utils.py:
```diff
@@ -1095,3 +1100,72 @@ def load_configuration(path: Union[str, pathlib.Path, os.PathLike]) -> Mapping[s
+    :param input_tensor: An input tensor with final dimension $d$.
+    """
+    step1_tensor = torch.abs(input_tensor)  # Compute absolute value of all entries
+    step2_tensor = step1_tensor + SANITY_EPSILON  # Prevent zero values by adding a sanity epsilon
```
```suggestion
    step2_tensor = step1_tensor.clamp(min=SANITY_EPSILON)  # Prevent zero values by adding a sanity epsilon
```
I usually prefer to clamp instead of adding a small value, since it leaves sufficiently large values completely unmodified. (although this might numerically play only a neglicible role )

ralphabb replied:
Fair enough, thanks :)

mberr commented on src/pykeen/utils.py:
```diff
@@ -112,6 +115,8 @@
 
 _CUDA_NONZERO_ERROR = "nonzero is not supported for tensors with more than INT_MAX elements"
 
+SANITY_EPSILON = 10 ** -8
```
I wonder whether we have this constant already somewhere?

I might also be good to use [torch.finfo(x.dtype).eps](https://pytorch.org/docs/stable/type_info.html#torch.torch.finfo) instead since it allows changing the default precision (e.g., for automatic mixed precision) and automatically updating these constants.

cthoyt replied:
TODO

cthoyt replied:
```suggestion
SANITY_EPSILON = torch.finfo(x.dtype).eps
```

mberr replied:
cf. #626

cthoyt replied:
Nice.

mberr commented on src/pykeen/utils.py:
```diff
@@ -1095,3 +1100,72 @@ def load_configuration(path: Union[str, pathlib.Path, os.PathLike]) -> Mapping[s
+    :param delta: The base shape of the input relation embeddings.
+    :param size: The size scalar vectors of the input relation embeddings.
+    """
+    size_pos = torch.nn.functional.elu(size) + 1  # Enforce that sizes are strictly positive by passing through ELU
```
Interesting choice: is there any reason why [ELU](https://pytorch.org/docs/stable/generated/torch.nn.ELU.html#torch.nn.ELU) + 1 is preferred over [softplus](https://pytorch.org/docs/stable/generated/torch.nn.Softplus.html#torch.nn.Softplus)?

ralphabb replied:
I guess the main motivation was that the value is simply x+1 when x is positive, but since we just need the size to be positive, softplus would also work.

mberr commented on src/pykeen/utils.py:
```diff
@@ -1095,3 +1100,72 @@ def load_configuration(path: Union[str, pathlib.Path, os.PathLike]) -> Mapping[s
+    """
+    size_pos = torch.nn.functional.elu(size) + 1  # Enforce that sizes are strictly positive by passing through ELU
+    delta_norm = product_normalise(delta)  # Shape vector is normalized using the above helper function
+    delta_final = torch.multiply(size_pos, delta_norm)  # Size is learned separately and applied to normalized shape
```
```suggestion
    delta_final = size_pos * delta_norm  # Size is learned separately and applied to normalized shape
```

mberr commented on src/pykeen/utils.py:
```diff
@@ -1095,3 +1100,72 @@ def load_configuration(path: Union[str, pathlib.Path, os.PathLike]) -> Mapping[s
+    :param base: The base position (box center) of the input relation embeddings.
+    :param delta: The base shape of the input relation embeddings.
+    :param size: The size scalar vectors of the input relation embeddings.
+    """
```
missing return value documentation

ralphabb replied:
Have now added this for all util functions related to BoxE

mberr commented on src/pykeen/utils.py:
```diff
@@ -1095,3 +1100,72 @@ def load_configuration(path: Union[str, pathlib.Path, os.PathLike]) -> Mapping[s
+    r"""Given sets of embeddings of base position, shape, and size, compute the lower and upper corners of
+    the resulting box
+
+    :param base: The base position (box center) of the input relation embeddings.
```
are there any constraints to the shape of these tensors? Or is the whole method working element-wise?

ralphabb replied:
So, in the BoxE code, the tensor there is Rx arity x d (so Rx2xd for the KG case), where R is the number of relations, but any tensor of shape (..., ..., ..,., d) should work here, so long as the shape is consistent across all inputs. The method is not element-wise only in that it includes product normalisation (line 1130).

cthoyt replied:
Please move this explanation into the docs

mberr commented on src/pykeen/nn/modules.py:
```diff
@@ -1418,3 +1420,66 @@ def _prepare_hrt_for_functional(
+        # Fifth, compute the norm
+        score_h = dist_h.norm(p=self.norm_order, dim=-1)
+        score_t = dist_t.norm(p=self.norm_order, dim=-1)
+        total_score = score_h + score_t
```
Can we reduce code duplication by having something like

```python
total_score = sum(
  do_something_for_one_box_and_h_or_t(pos, bump, base, delta, size)
  for pos, bump, base, delta, size in (
    h + r[0:3],
    t + r[3:6],
  )
)
```
? Without spending too much time into deeply inspecting the code, it looks pretty parallel for h/t and the respective box.

ralphabb replied:
It is quite parallel indeed. I will write a helper function for this and put it in utils.

mberr reviewed (commented):
I also gave the code a quick review.

mberr committed:
Merge branch 'master' into master

ralphabb committed:
return doc for BoxE utility functions

cthoyt committed:
Add default loss function

cthoyt committed:
Update boxe_kg.py

cthoyt committed:
Add reported values from paper

I used the BoxE(a) instead of BoxE(u) but didn't look into the paper to see what the difference was

ralphabb committed:
Code refactoring

I made the position-wise computation modular for better readability.

cthoyt committed:
Borrow update from #623

mberr committed:
Merge branch 'master' into master

ralphabb committed:
Tester.py using mberr logger

ralphabb committed:
Merge branch 'master' of github.com:ralphabb/pykeen

mberr commented on Tester.py:
```diff
@@ -0,0 +1,79 @@
+        relation_initializer=torch.nn.init.uniform_,
+        relation_initializer_kwargs=init_kw,
+        relation_size_initializer=torch.nn.init.uniform_,
+        relation_size_initializer_kwargs=size_init_kw,
```
@ralphabb 
since you initialize the model outside of the pipeline, you need to add the loss here instead of providing it via the pipeline.

Alternatively, you can set

```python
pipeline(
  model=BoxEKG,
  model_kwargs=dict(
        embedding_dim=500,
        norm_order=2,
        tanh_map=True,
        entity_initializer=torch.nn.init.uniform_,
        entity_initializer_kwargs=init_kw,
        relation_initializer=torch.nn.init.uniform_,
        relation_initializer_kwargs=init_kw,
        relation_size_initializer=torch.nn.init.uniform_,
        relation_size_initializer_kwargs=size_init_kw,
  ),
)
```

ralphabb committed:
BoxE init now done in pipeline.

mberr committed:
add assertion

mberr committed:
fix NSSA + SLCWA training loop

trigger ci

cthoyt committed:
Merge remote-tracking branch 'origin/fix-nssa-loss-shape' into pr/618

ralphabb committed:
Merge branch 'master' of github.com:ralphabb/pykeen

cthoyt committed:
Merge branch 'master' into master

cthoyt committed:
Add imports

cthoyt committed:
Update README.md

cthoyt committed:
Add citation

cthoyt committed:
Apply linting

cthoyt committed:
Update src/pykeen/utils.py

Co-authored-by: Max Berrendorf <berrendorf@dbs.ifi.lmu.de>

cthoyt committed:
Reorganize and rename

cthoyt committed:
Update src/pykeen/utils.py

Co-authored-by: Max Berrendorf <berrendorf@dbs.ifi.lmu.de>

cthoyt committed:
Improve docs

cthoyt committed:
Update docs

cthoyt committed:
Rename model

mberr commented on src/pykeen/experiments/boxe/abboud2020_boxe_fb15k37.yaml:
```diff
@@ -0,0 +1,10 @@
+metadata:
```
TODO: if we want the experimental configuration to be part of this PR, we should fill this one out, too.

cthoyt replied:
yes I was getting there! Thank you :)

mberr commented on src/pykeen/models/unimodal/boxe.py:
```diff
@@ -0,0 +1,132 @@
+            restricts the embedding space to the range [-1, 1], and thus this map implicitly
+            regularizes the space to prevent loss reduction by growing boxes arbitrarily large.
+        :param norm_order:
+            Norm Order in score computation (Int): Default - 2
```
```suggestion
            Norm Order in score computation
```
The default value & type are already part of the signature, so we do not need to repeat them.

mberr commented on src/pykeen/models/unimodal/boxe.py:
```diff
@@ -0,0 +1,132 @@
+    def __init__(
+        self,
+        *,
+        embedding_dim: int = 200,
```
Is there any specific reason why this is 200? If not, I would prefer a embedding dimension which allows utilization of tensor cores (multiple of 8, larger powers of two are even better for performance)

```suggestion
        embedding_dim: int = 256,
```

ralphabb replied:
The main reason the default is 200 is because this is the standard value in the literature on e.g., YAGO3-10 and higher-arity benchmarks. No problem with changing it.

mberr replied:
Your argument is also quite good :confused: 

I think, I would still change it, since we have the experiment configurations (the YAMLs mentioned below) which explicitly provide all settings to re-run a certain experiment.

cthoyt committed:
Update docs

sorry but also have to make english usage standard USA style

cthoyt committed:
Delete Tester.py

This is now in the PR description

mberr commented on src/pykeen/nn/modules.py:
```diff
@@ -1413,6 +1416,64 @@ def _prepare_hrt_for_functional(
+        self.tanh_map = tanh_map  # Map the tanh map
+        self.norm_order = norm_order
+
+    def forward(self, h, r, t):
```
missing type annotation (cf. super class).

also add `# noqa: D102` for flake8

mberr commented on src/pykeen/nn/modules.py:
```diff
@@ -1413,6 +1416,64 @@ def _prepare_hrt_for_functional(
+            norm_order=self.norm_order,
+        )
+        total_score = score_h + score_t
+        return -total_score  # Because this is inverted in NSSALoss (higher is better)
```
In general, we follow the convention that larger scores indicate more likely triples. Hence, for distance based models, we usually use `score = -distance` (although of course other anti-monotonous transformation are also possible).

ralphabb replied:
Yes, I inferred this. All good :)

cthoyt replied:
~Can you make this shorter by just doing `score_t - score_h` and adding a comment that you flipped it to make it negative. Then you save one operation~

nope nope nope that is not correct

mberr replied:
>Can you make this shorter by just doing score_t - score_h and adding a comment that you flipped it to make it negative. Then you save one operation

would be `-score_h - score_t`, right? In that case, no operation was saved

cthoyt replied:
Yes. You are right. I have now lost the skill to do basic algebra. At first I just couldn't do arithmetic. Soon I will forget calculus as well, I guess

mberr commented on src/pykeen/utils.py:
```diff
@@ -1095,3 +1100,118 @@ def load_configuration(path: Union[str, pathlib.Path, os.PathLike]) -> Mapping[s
+    step3_tensor = torch.mean(log_norm_tensor, dim=-1, keepdim=True)
+    norm_volume = torch.exp(step3_tensor)
+    pre_norm_out = input_tensor / norm_volume
+    return pre_norm_out
```
```suggestion
    return input_tensor / (input_tensor.abs()\
       .clamp(min=SANITY_EPSILON)\
       .log()\
       .mean(dim=1, keepdim=True)\
       .exp()\
       .clamp(min=SANITY_EPSILON))
```

not super beautiful, but does not keep references to intermediate tensors, and thus allows earlier garbage collection / memory re-use.

I also added clamping before division - I think this should not be necessary (since `exp() > 0`), but just to be sure for numeric instability.

ralphabb replied:
All good! So long as it's equivalent :)

cthoyt replied:
Isn't `input_tensor / input_tensor.abs()` also the same as taking the norm with `p=1`? Then you can achieve maximum chainage

mberr replied:
> Isn't input_tensor / input_tensor.abs() also the same as taking the norm with p=1? Then you can achieve maximum chainage

no, the full right hand side is evaluated before the division. I'll add a bracket to make this clearer :sweat_smile:

cthoyt replied:
just let black do what it wants and add a code comment to prevent future non-careful readers like me ;)

cthoyt replied:
Added in d7264a4a

cthoyt committed:
Update modules.py

mberr commented on src/pykeen/utils.py:
```diff
@@ -1095,3 +1100,118 @@ def load_configuration(path: Union[str, pathlib.Path, os.PathLike]) -> Mapping[s
+
+
+def compute_box(
+    base: torch.FloatTensor, delta: torch.FloatTensor, size: torch.FloatTensor
```
trailing comma

cthoyt committed:
Add typing

mberr commented on src/pykeen/utils.py:
```diff
@@ -1095,3 +1100,118 @@ def load_configuration(path: Union[str, pathlib.Path, os.PathLike]) -> Mapping[s
+
+
+def point_to_box_distance(
+    points: torch.FloatTensor, box_lows: torch.FloatTensor, box_highs: torch.FloatTensor
```
trailing comma

cthoyt replied:
it is the will of `black` to not include trailing comma here

mberr replied:
You can add it, and it will not remove it, but introduce the line breaks we see at many other places :confused:

mberr commented on src/pykeen/utils.py:
```diff
@@ -1095,3 +1100,118 @@ def load_configuration(path: Union[str, pathlib.Path, os.PathLike]) -> Mapping[s
+
+    .. math::
+
+        dist(p,l,h) = \begin{cases}
```
@cthoyt can we use `cases` in docstrings?

cthoyt replied:
yup. I just built it
<img width="835" alt="Screen Shot 2021-11-09 at 12 43 39" src="https://user-images.githubusercontent.com/5069736/140918196-eb5e9cba-cc29-43f1-a3ae-b5fb832a2b93.png">

mberr commented on src/pykeen/utils.py:
```diff
@@ -1095,3 +1100,118 @@ def load_configuration(path: Union[str, pathlib.Path, os.PathLike]) -> Mapping[s
+        bumped_representation = torch.tanh(bumped_representation)
+    # Compute the distance function output element-wise
+    element_wise_distance = point_to_box_distance(
+        points=bumped_representation, box_lows=relation_box_low, box_highs=relation_box_high
```
trailing comma

mberr commented on src/pykeen/utils.py:
```diff
@@ -1095,3 +1100,118 @@ def load_configuration(path: Union[str, pathlib.Path, os.PathLike]) -> Mapping[s
+        points=bumped_representation, box_lows=relation_box_low, box_highs=relation_box_high
+    )
+    # Finally, compute the norm
+    overall_score = element_wise_distance.norm(p=norm_order, dim=-1)
```
for translational models (e.g. TransE), we also support power norm, i..e, `|x|_p^p`, which has some nice numeric properties (and was used by some other interaction models, too). Can we also support this here?

cthoyt replied:
![](https://c.tenor.com/7EttrztgjiAAAAAM/palpatine-star-wars.gif)

ralphabb replied:
I haven't experimented with this as part of BoxE, but I don't see a reason why this wouldn't be supported for aggregating distance scores.

cthoyt replied:
@mberr can take care of this one directly

mberr replied:
ab6deac

mberr commented on src/pykeen/utils.py:
```diff
@@ -1095,3 +1100,118 @@ def load_configuration(path: Union[str, pathlib.Path, os.PathLike]) -> Mapping[s
+    )
+    # Finally, compute the norm
+    overall_score = element_wise_distance.norm(p=norm_order, dim=-1)
+    return overall_score
```
```suggestion
    return element_wise_distance.norm(p=norm_order, dim=-1)
```

the variable is not used anymore, and directly returned afterwards.

mberr reviewed (commented):
I reviewed once more.

Trailing commas are also something I can take care of, but I did not want to commit stylistic changes to your branch in case there are any uncommitted / unpushed changes that may conflict.

cthoyt committed:
Update docs

mberr commented on src/pykeen/experiments/boxe/abboud2020_boxe_fb15k37.yaml:
```diff
@@ -0,0 +1,10 @@
+  title: Learn FB15k-237 Dataset with BoxE as described by Abboud et al., 2020
+results:
+  hits_at_k:
+    FIXME:
```
It *seems* to be optimistic, although I am not 100% certain about it.

https://github.com/ralphabb/BoxE/blob/584c83bb42817c67df0f70c52898aacfc916af19/BoxEModel.py#L915-L925

Maybe @ralphabb can help here :slightly_smiling_face:

cthoyt replied:
I think we can leave it to @ralphabb to curate his results ;) 

just follow the same form as the other examples in the src/pykeen/experiments directory

ralphabb replied:
I'm afraid I don't understand what you mean here. Can you re-phrase this?

cthoyt replied:
When we implement new models in PyKEEN, we typically curate the experimental configurations reported in their accompanying papers. For example, the R-GCN paper reported results for FB15k, so we curated the configuration (i.e., the loss used, regularization used, hyperparameters, etc.) [here](https://github.com/pykeen/pykeen/blob/master/src/pykeen/experiments/rgcn/schlichtkrull2018_rgcn_fb15k.yaml). This YAML file corresponds to the arguments that would go into the `pykeen.pipeline.pipeline()` function to directly run the same experiment to try and reproduce the results reported in the paper. 

Note that the R-GCN configuration is pretty complicated, the BoxE one should be much less!

If they did HPO, we pick the configuration for the best run, since usually papers report the best run in their main text

ralphabb replied:
Oh so should just pass the configurations (kwargs) that align with my repo setups?

cthoyt replied:
Yup! let us know if you have any questions about how to write it

ralphabb replied:
I've written them. I'm not very sure if uniform_norm does the job (I was using uniform_ in Tester), and I couldn't find a pointer in the docs to validate the range it produces. In any case, I have clarified the intended init as a comment for us to change if this isn't currently correct.

cthoyt committed:
Update src/pykeen/utils.py

Co-authored-by: Max Berrendorf <berrendorf@dbs.ifi.lmu.de>

cthoyt reviewed (changes requested):
Here's the TODO list before we can merge this:

1. Add experimental configurations for best runs from the BoxE paper
2. Address all of Max's code comments
3. Run CI/CD

I think all of the code is now in the right shape, thanks so much @ralphabb for going back and forth with us until now. This has been lots of fun :)

mberr committed:
Change default value for embedding_dim

This makes the model by default better suited for 
current GPU architectures

ralphabb committed:
Merge branch 'master' of github.com:ralphabb/pykeen

ralphabb committed:
Merge branch 'master' of github.com:ralphabb/pykeen

cthoyt committed:
Update src/pykeen/models/unimodal/boxe.py

Co-authored-by: Max Berrendorf <berrendorf@dbs.ifi.lmu.de>

cthoyt committed:
Update utils.py

mberr committed:
add functional form & power norm support

mberr committed:
Merge branch 'master' of github.com:ralphabb/pykeen into pr/ralphabb/618

mberr committed:
reformat

mberr committed:
update BoxE to changed parameter name

cthoyt commented on src/pykeen/nn/functional.py:
```diff
@@ -1186,3 +1189,49 @@ def cross_e_interaction(
+    This interaction relies on Abboud's point-to-box distance
+    :func:`pykeen.utils.point_to_box_distance`.
+    """
+    # TODO: Docstring
```
TODO: docstring should very explicitly explain what each of the embeddings does / why

cthoyt commented on src/pykeen/nn/functional.py:
```diff
@@ -1186,3 +1189,49 @@ def cross_e_interaction(
+    rt_low, rt_high = compute_box(rt_base, rt_delta, rt_size)
+
+    return sum(
+        boxe_kg_arity_position_computation(  # ~ negative distance
```
I'd rather the negative sum of norms happens in this function than inside `boxe_kg_arity_position_computation()`

mberr replied:
I would prefer that too, but this would require to either have a new method, or to update all other places where this is used.

mberr replied:
Maybe we can update the helper method name / description to describe that this already returns a score (in our normal definition, i.e., larger is better)

cthoyt commented on src/pykeen/nn/modules.py:
```diff
@@ -1413,6 +1414,65 @@ def _prepare_hrt_for_functional(
+            rt_base=rt_base,
+            rt_delta=rt_delta,
+            rt_size=rt_size,
+            #
```
don't forget comments for these two bits

cthoyt reviewed (commented):
added some more comments, github is being picky about grouping them as a "review" now for some reason

mberr committed:
update docstring

mberr committed:
reformat and update docstring

cthoyt commented on src/pykeen/utils.py:
```diff
@@ -1095,3 +1100,144 @@ def load_configuration(path: Union[str, pathlib.Path, os.PathLike]) -> Mapping[s
+    second_bound = base + 0.5 * delta_final
+
+    # Compute box upper bounds using min and max respectively
+    box_low = torch.minimum(first_bound, second_bound)
```
would it be more simple just to check if base < 0 and assign these in the conditional?

mberr replied:
No, since this operation happens in parallel / vectorized, whereas a "normal" conditional would either require conversion from torch to Python (bad, in particular since the needs to be transferred to cpu!), or usage of torch.where, but this should be roughly the same speed.

cthoyt replied:
cool!

mberr committed:
update point_to_box_distance

mberr committed:
update docstring

mberr committed:
move main to end of file

mberr committed:
Add BoxE functional docstring

mberr committed:
rename method and update docstring

cthoyt committed:
Merge branch 'master' into pr/618

mberr committed:
move box unpacking into method

mberr committed:
Merge branch 'master' of github.com:ralphabb/pykeen into pr/ralphabb/618

cthoyt committed:
Use `at_least_eps()` for clamping

cthoyt committed:
Fix typo

mberr committed:
update docstring

trigger ci

mberr committed:
Merge branch 'master' of github.com:ralphabb/pykeen into pr/ralphabb/618

cthoyt commented on src/pykeen/nn/functional.py:
```diff
@@ -1186,3 +1189,82 @@ def cross_e_interaction(
+    :param rh_base: shape: (batch_size, 1, num_relations, 1, d)
+        the relation-specific head box base position
+    :param rh_delta: shape: (batch_size, 1, num_relations, 1, d)
+        # TODO:
```
TODO

ralphabb replied:
On it

cthoyt commented on src/pykeen/nn/functional.py:
```diff
@@ -1186,3 +1189,82 @@ def cross_e_interaction(
+    :param rt_base: shape: (batch_size, 1, num_relations, 1, d)
+        the relation-specific tail box base position
+    :param rt_delta: shape: (batch_size, 1, num_relations, 1, d)
+        # TODO:
```
TODO

ralphabb replied:
On it

cthoyt reviewed (commented):
See TODOs

ralphabb committed:
Merge branch 'master' of github.com:ralphabb/pykeen

ralphabb committed:
Experimental Setups curated

I added all the setups. Not very sure about uniform_norm. Have commented what I intend to achieve here.

ralphabb committed:
norm_order replaced with p

cthoyt committed:
Cleanup YAML

cthoyt committed:
Fix indentation

cthoyt committed:
Add missing flag for power norm

cthoyt committed:
Pass flake8

Trigger CI

cthoyt committed:
Merge branch 'master' into pr/618

ralphabb committed:
Loss comment in YAML updated

ralphabb committed:
Merge branch 'master' of github.com:ralphabb/pykeen

PyKEEN-bot committed:
Trigger CI

ralphabb committed:
Merge branch 'master' of github.com:ralphabb/pykeen

PyKEEN-bot committed:
Trigger CI

ralphabb committed:
Note about BoxE higher-arity and reference to complete implementation.

ralphabb committed:
Merge branch 'master' of github.com:ralphabb/pykeen

cthoyt committed:
Update boxe.py

PyKEEN-bot committed:
Trigger CI

cthoyt committed:
Update docs

PyKEEN-bot committed:
Trigger CI

mberr commented on src/pykeen/models/unimodal/boxe.py:
```diff
@@ -0,0 +1,149 @@
+        self,
+        *,
+        embedding_dim: int = 256,
+        tanh_map: bool = True,
```
1. tanh_map comes from here

mberr commented on src/pykeen/models/unimodal/boxe.py:
```diff
@@ -0,0 +1,149 @@
+            interaction_kwargs=dict(
+                p=p,
+                power_norm=power_norm,
+                tanh_map=tanh_map,
```
2. goes into here

mberr commented on src/pykeen/nn/modules.py:
```diff
@@ -1413,6 +1414,65 @@ def _prepare_hrt_for_functional(
+    relation_shape = ("d", "d", "s", "d", "d", "s")  # Boxes are 2xd (size) each, x 2 sets of boxes: head and tail
+    entity_shape = ("d", "d")  # Base position and bump
+
+    def __init__(self, tanh_map: bool = True, p: int = 2, power_norm: bool = False):
```
3. and then here

mberr commented on src/pykeen/nn/modules.py:
```diff
@@ -1413,6 +1414,65 @@ def _prepare_hrt_for_functional(
+            whether to use the p-th power of the norm instead
+        """
+        super().__init__(p=p, power_norm=power_norm)
+        self.tanh_map = tanh_map
```
4. and here

mberr commented on src/pykeen/nn/modules.py:
```diff
@@ -1413,6 +1414,65 @@ def _prepare_hrt_for_functional(
+
+    def _prepare_state_for_functional(self) -> MutableMapping[str, Any]:  # noqa: D102
+        state = super()._prepare_state_for_functional()
+        state["tanh_map"] = self.tanh_map
```
5. here it should go into the kwargs for the functional version

mberr commented on src/pykeen/nn/functional.py:
```diff
@@ -1186,3 +1189,82 @@ def cross_e_interaction(
+    t_pos: torch.FloatTensor,
+    t_bump: torch.FloatTensor,
+    # power norm
+    tanh_map: bool = False,
```
6. and thus arrive here :confused:

ralphabb replied:
I don't know, really. All I can say, very confidently, is that the model works when I replace this False with True.

ralphabb replied:
This should be changed, anyway, but it suggests a problem if the default is being used in the first place.

mberr reviewed (commented):
I have not found the reason why the parameter is not passed from the interaction module.

mberr commented on src/pykeen/nn/modules.py:
```diff
@@ -1413,6 +1414,65 @@ def _prepare_hrt_for_functional(
+        super().__init__(p=p, power_norm=power_norm)
+        self.tanh_map = tanh_map
+
+    def _prepare_for_functional(
```
got it. we are overwriting the wrong function!

This should be `_prepare_hrt_for_functional`.

cthoyt replied:
^^^^

hmm guess who skipped the unit tests for the BoxEInteracton module

mberr replied:
I am not sure whether we would detect this in our tests - is there a way how to enforce that subclasses implement certain methods? (is `@abstractmethod` sufficient, or is this just some annotation for linters?)
